File: synapse/zig/src/ops/batchnorm.zig
//! Batch normalization with Welford single-pass mean+variance.
//! Supports training mode (compute batch stats, update running stats)
//! and inference mode (use running stats).

const std = @import("std");
const shape_mod = @import("../tensor/shape.zig");
const tensor_mod = @import("../tensor/tensor.zig");
const storage_mod = @import("../tensor/storage.zig");

const Shape = shape_mod.Shape;
const Tensor = tensor_mod.Tensor;
const Storage = storage_mod.Storage;

pub const BatchNorm = struct {
    num_features: usize,
    eps: f32,
    momentum: f32,
    gamma: []f32,
    beta: []f32,
    running_mean: []f32,
    running_var: []f32,
    allocator: std.mem.Allocator,

    /// Create a BatchNorm layer for `num_features` channels.
    /// gamma initialized to 1, beta to 0, running_mean to 0, running_var to 1.
    pub fn init(allocator: std.mem.Allocator, num_features: usize, eps: f32, momentum: f32) !BatchNorm {
        const gamma = try allocator.alloc(f32, num_features);
        for (gamma) |*g| g.* = 1.0;

        const beta = try allocator.alloc(f32, num_features);
        for (beta) |*b| b.* = 0.0;

        const running_mean = try allocator.alloc(f32, num_features);
        for (running_mean) |*m| m.* = 0.0;

        const running_var = try allocator.alloc(f32, num_features);
        for (running_var) |*v| v.* = 1.0;

        return .{
            .num_features = num_features,
            .eps = eps,
            .momentum = momentum,
            .gamma = gamma,
            .beta = beta,
            .running_mean = running_mean,
            .running_var = running_var,
            .allocator = allocator,
        };
    }

    /// Free all owned buffers.
    pub fn deinit(self: *BatchNorm) void {
        self.allocator.free(self.gamma);
        self.allocator.free(self.beta);
        self.allocator.free(self.running_mean);
        self.allocator.free(self.running_var);
    }

    /// Forward pass. Input shape: [N, C] where C == num_features.
    /// In training mode, computes batch mean/var with Welford and updates running stats.
    /// In inference mode, uses running stats.
    pub fn forward(self: *BatchNorm, allocator: std.mem.Allocator, input: Tensor(f32), training: bool) !Tensor(f32) {
        std.debug.assert(input.shape.ndim == 2);
        const batch_size = input.shape.dims[0];
        const num_channels = input.shape.dims[1];
        std.debug.assert(num_channels == self.num_features);

        const out_storage = try Storage.create(allocator, f32, input.numel());
        const out = Tensor(f32).init(out_storage, input.shape);
        out_storage.release();

        const in_data = input.storage.dataAs(f32);
        const out_data = out.storage.dataAs(f32);
        const in_stride0 = input.strides[0];
        const in_stride1 = input.strides[1];

        for (0..num_channels) |c| {
            var used_mean: f32 = undefined;
            var used_var: f32 = undefined;

            if (training) {
                // Welford single-pass mean + variance.
                const result = welfordMeanVar(in_data, input.offset, in_stride0, in_stride1, c, batch_size);
                used_mean = result.mean;
                used_var = result.variance;

                // Update running stats with exponential moving average.
                self.running_mean[c] = (1.0 - self.momentum) * self.running_mean[c] + self.momentum * used_mean;
                self.running_var[c] = (1.0 - self.momentum) * self.running_var[c] + self.momentum * used_var;
            } else {
                used_mean = self.running_mean[c];
                used_var = self.running_var[c];
            }

            const inv_std = 1.0 / @sqrt(used_var + self.eps);
            const g = self.gamma[c];
            const b = self.beta[c];

            for (0..batch_size) |n| {
                const in_idx = input.offset + n * in_stride0 + c * in_stride1;
                const out_idx = n * num_channels + c;
                out_data[out_idx] = g * (in_data[in_idx] - used_mean) * inv_std + b;
            }
        }

        return out;
    }
};

/// Welford single-pass mean + variance for a column of a 2D tensor.
pub fn welfordMeanVar(
    data: []f32,
    offset: usize,
    stride0: usize,
    stride1: usize,
    channel: usize,
    batch_size: usize,
) struct { mean: f32, variance: f32 } {
    var mean: f32 = 0;
    var m2: f32 = 0;

    for (0..batch_size) |n| {
        const x = data[offset + n * stride0 + channel * stride1];
        const count_f: f32 = @floatFromInt(n + 1);
        const delta = x - mean;
        mean += delta / count_f;
        const delta2 = x - mean;
        m2 += delta * delta2;
    }

    const batch_f: f32 = @floatFromInt(batch_size);
    return .{ .mean = mean, .variance = m2 / batch_f };
}

/// Two-pass mean + variance (for benchmark comparison).
pub fn twoPassMeanVar(
    data: []f32,
    offset: usize,
    stride0: usize,
    stride1: usize,
    channel: usize,
    batch_size: usize,
) struct { mean: f32, variance: f32 } {
    const batch_f: f32 = @floatFromInt(batch_size);

    // Pass 1: mean
    var sum: f32 = 0;
    for (0..batch_size) |n| {
        sum += data[offset + n * stride0 + channel * stride1];
    }
    const mean = sum / batch_f;

    // Pass 2: variance
    var sq_sum: f32 = 0;
    for (0..batch_size) |n| {
        const diff = data[offset + n * stride0 + channel * stride1] - mean;
        sq_sum += diff * diff;
    }

    return .{ .mean = mean, .variance = sq_sum / batch_f };
}

File: synapse/crates/synapse-nn/src/batchnorm.rs
//! Batch normalization layers: BatchNorm1d, BatchNorm2d.

use synapse_autograd::Tensor;

use crate::module::Module;

// ── BatchNorm1d ───────────────────────────────────────────────────────

/// Batch normalization over a 2D input [N, C] or 3D input [N, C, L].
/// Normalizes over the batch (and spatial) dimensions, per-channel.
pub struct BatchNorm1d {
    pub num_features: usize,
    pub gamma: Tensor,        // scale, shape [C]
    pub beta: Tensor,         // shift, shape [C]
    pub running_mean: Tensor, // [C]
    pub running_var: Tensor,  // [C]
    pub eps: f32,
    pub momentum: f32,
    training: bool,
}

impl BatchNorm1d {
    pub fn new(num_features: usize) -> Self {
        BatchNorm1d {
            num_features,
            gamma: Tensor::ones(&[num_features]),
            beta: Tensor::zeros(&[num_features]),
            running_mean: Tensor::zeros(&[num_features]),
            running_var: Tensor::ones(&[num_features]),
            eps: 1e-5,
            momentum: 0.1,
            training: true,
        }
    }
}

impl Module for BatchNorm1d {
    /// Forward: input [N, C] or [N, C, L] -> same shape
    fn forward(&self, input: &Tensor) -> Tensor {
        let ndim = input.shape.len();
        assert!(ndim == 2 || ndim == 3, "BatchNorm1d expects 2D or 3D input");
        let c = input.shape[1];
        assert_eq!(c, self.num_features);
        let batch = input.shape[0];
        let spatial: usize = if ndim == 3 { input.shape[2] } else { 1 };

        let mut output = vec![0.0f32; input.numel()];

        for ci in 0..c {
            let (mean, var) = if self.training {
                // Compute mean and variance over batch and spatial dims
                let mut sum = 0.0f32;
                let count = (batch * spatial) as f32;
                for n in 0..batch {
                    for s in 0..spatial {
                        let idx = n * c * spatial + ci * spatial + s;
                        sum += input.data[idx];
                    }
                }
                let mean = sum / count;
                let mut var_sum = 0.0f32;
                for n in 0..batch {
                    for s in 0..spatial {
                        let idx = n * c * spatial + ci * spatial + s;
                        let diff = input.data[idx] - mean;
                        var_sum += diff * diff;
                    }
                }
                let var = var_sum / count;
                (mean, var)
            } else {
                (self.running_mean.data[ci], self.running_var.data[ci])
            };

            let inv_std = 1.0 / (var + self.eps).sqrt();
            let g = self.gamma.data[ci];
            let b = self.beta.data[ci];

            for n in 0..batch {
                for s in 0..spatial {
                    let idx = n * c * spatial + ci * spatial + s;
                    output[idx] = g * (input.data[idx] - mean) * inv_std + b;
                }
            }
        }

        Tensor::new(output, input.shape.clone())
    }

    fn parameters(&self) -> Vec<&Tensor> {
        vec![&self.gamma, &self.beta]
    }

    fn parameters_mut(&mut self) -> Vec<&mut Tensor> {
        vec![&mut self.gamma, &mut self.beta]
    }

    fn set_training(&mut self, training: bool) {
        self.training = training;
    }

    fn is_training(&self) -> bool {
        self.training
    }

    fn name(&self) -> &str {
        "BatchNorm1d"
    }
}

// ── BatchNorm2d ───────────────────────────────────────────────────────

/// Batch normalization over a 4D input [N, C, H, W].
/// Normalizes over N, H, W dimensions per channel.
pub struct BatchNorm2d {
    pub num_features: usize,
    pub gamma: Tensor,        // [C]
    pub beta: Tensor,         // [C]
    pub running_mean: Tensor, // [C]
    pub running_var: Tensor,  // [C]
    pub eps: f32,
    pub momentum: f32,
    training: bool,
}

impl BatchNorm2d {
    pub fn new(num_features: usize) -> Self {
        BatchNorm2d {
            num_features,
            gamma: Tensor::ones(&[num_features]),
            beta: Tensor::zeros(&[num_features]),
            running_mean: Tensor::zeros(&[num_features]),
            running_var: Tensor::ones(&[num_features]),
            eps: 1e-5,
            momentum: 0.1,
            training: true,
        }
    }
}

impl Module for BatchNorm2d {
    /// Forward: input [N, C, H, W] -> same shape
    fn forward(&self, input: &Tensor) -> Tensor {
        assert_eq!(
            input.shape.len(),
            4,
            "BatchNorm2d expects 4D input [N, C, H, W]"
        );
        let batch = input.shape[0];
        let c = input.shape[1];
        let h = input.shape[2];
        let w = input.shape[3];
        assert_eq!(c, self.num_features);

        let hw = h * w;
        let mut output = vec![0.0f32; input.numel()];

        for ci in 0..c {
            let (mean, var) = if self.training {
                let count = (batch * hw) as f32;
                let mut sum = 0.0f32;
                for n in 0..batch {
                    let base = n * c * hw + ci * hw;
                    for j in 0..hw {
                        sum += input.data[base + j];
                    }
                }
                let mean = sum / count;
                let mut var_sum = 0.0f32;
                for n in 0..batch {
                    let base = n * c * hw + ci * hw;
                    for j in 0..hw {
                        let diff = input.data[base + j] - mean;
                        var_sum += diff * diff;
                    }
                }
                let var = var_sum / count;
                (mean, var)
            } else {
                (self.running_mean.data[ci], self.running_var.data[ci])
            };

            let inv_std = 1.0 / (var + self.eps).sqrt();
            let g = self.gamma.data[ci];
            let b = self.beta.data[ci];

            for n in 0..batch {
                let base = n * c * hw + ci * hw;
                for j in 0..hw {
                    output[base + j] = g * (input.data[base + j] - mean) * inv_std + b;
                }
            }
        }

        Tensor::new(output, input.shape.clone())
    }

    fn parameters(&self) -> Vec<&Tensor> {
        vec![&self.gamma, &self.beta]
    }

    fn parameters_mut(&mut self) -> Vec<&mut Tensor> {
        vec![&mut self.gamma, &mut self.beta]
    }

    fn set_training(&mut self, training: bool) {
        self.training = training;
    }

    fn is_training(&self) -> bool {
        self.training
    }

    fn name(&self) -> &str {
        "BatchNorm2d"
    }
}
